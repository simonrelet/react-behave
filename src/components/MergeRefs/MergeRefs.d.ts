import * as React from 'react'

export interface RefFunction {
  (ref: HTMLElement): void
}

export interface RenderProps {
  ref: RefFunction
}

export interface MergeRefsProps {
  children(props: RenderProps): React.ReactNode
  refs: Array<RefFunction | React.RefObject<HTMLElement>>
}

export declare const MergeRefs: React.ComponentType<MergeRefsProps>
