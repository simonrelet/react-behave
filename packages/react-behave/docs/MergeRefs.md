# MergeRefs

[callback-refs]: https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
[create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref

> Merge multiple refs on a single HTML element.

## Usage

```jsx
import React, { Component } from 'react';
import { ComponentNeedRef } from 'some-lib';
import { MergeRefs } from 'react-behave';

class App extends Component {
  buttonRef = React.createRef();

  componentDidMount() {
    this.buttonRef.focus();
  }

  render() {
    return (
      <ComponentNeedRef
        render={requiredRef => (
          <MergeRefs
            refs={[requiredRef, this.buttonRef]}
            render={ref => <button ref={ref}>Click me</button>}
          />
        )}
      />
    );
  }
}
```

This component is an alternative of wrapping `<div />`s when you need to have multiple refs on a single HTML element.

## Props

_Mandatory props are marked with a `*`._

| Name                                   |            Type             | Default value | Description                                                                                                                                              |
| -------------------------------------- | :-------------------------: | :-----------: | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <strong><code>refs</code>\*</strong>   | `arrayOf`<`func`\|`object`> |               | Array of refs. Each ref can either be a [callback ref][callback-refs] or an object created with [`React.createRef`][create-ref]. Falsy refs are ignored. |
| <strong><code>render</code>\*</strong> |           `func`            |               | Render function. The parameter is either a callback ref or `null` if `refs` is empty or only contain falsy values.                                       |
