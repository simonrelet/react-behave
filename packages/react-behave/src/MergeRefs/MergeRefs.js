import { setRef } from '@simonrelet/react-utils';
import PropTypes from 'prop-types';
import { Component } from 'react';

/**
 * [callback-refs]: https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
 * [create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref
 *
 * > Merge multiple refs on a single HTML element.
 *
 * ## Usage
 *
 * ```jsx
 * import React, { Component } from 'react';
 * import { ComponentNeedRef } from 'some-lib';
 * import { MergeRefs } from 'react-behave';
 *
 * class App extends Component {
 *   buttonRef = React.createRef();
 *
 *   componentDidMount() {
 *     this.buttonRef.focus();
 *   }
 *
 *   render() {
 *     return (
 *       <ComponentNeedRef
 *         render={requiredRef => (
 *           <MergeRefs
 *             refs={[requiredRef, this.buttonRef]}
 *             render={ref => <button ref={ref}>Click me</button>}
 *           />
 *         )}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * This component is an alternative of wrapping `<div />`s when you need to have multiple refs on a single HTML element.
 */
class MergeRefs extends Component {
  static propTypes = {
    /**
     * Array of refs.
     * Each ref can either be a [callback ref][callback-refs] or an object created with [`React.createRef`][create-ref].
     * Falsy refs are ignored.
     */
    refs: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    ).isRequired,

    /**
     * Render function.
     * The parameter is either a callback ref or `null` if `refs` is empty or only contain falsy values.
     */
    render: PropTypes.func.isRequired,
  };

  state = {
    refs: [],
  };

  static getDerivedStateFromProps(props, state) {
    return {
      refs: props.refs.filter(Boolean),
    };
  }

  mergeRefs = ref => {
    this.state.refs.forEach(handler => {
      setRef(ref, handler);
    });
  };

  render() {
    // Don't pass the ref function if there is no refs.
    return this.props.render(
      this.state.refs.length > 0 ? this.mergeRefs : null,
    );
  }
}

export default MergeRefs;
