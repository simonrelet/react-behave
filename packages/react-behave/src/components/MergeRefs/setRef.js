function setRef(ref, handler) {
  if (typeof handler === 'function') {
    handler(ref);
  } else if (typeof handler === 'object') {
    handler.current = ref;
  } else {
    throw new Error(
      [
        'Only refs of type `function` and `React.createRef()` are supported',
        `Got ${typeof handler}`,
      ].join('\n'),
    );
  }
}

export default setRef;
