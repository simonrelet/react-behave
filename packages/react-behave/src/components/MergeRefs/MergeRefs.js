import PropTypes from 'prop-types';
import { Component } from 'react';
import setRef from './setRef';

/**
 * [callback-refs]: https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
 * [create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref
 *
 * Merge multiple refs on a single component.
 *
 * ## Usage
 *
 * ```jsx
 * import React, { Component } from 'react';
 * import { ComponentNeedRef } from 'some-lib';
 * import { MergeRefs } from 'react-behave';
 *
 * class App extends Component {
 *   buttonRef = React.createRef();
 *
 *   componentDidMount() {
 *     this.buttonRef.focus();
 *   }
 *
 *   render() {
 *     return (
 *       <ComponentNeedRef
 *         render={requiredRef => (
 *           <MergeRefs
 *             refs={[requiredRef, this.buttonRef]}
 *             render={ref => <button ref={ref}>Click me</button>}
 *           />
 *         )}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * This component is an alternative of wrapping `<div />`s when you need to have multiple refs on a single component.
 */
class MergeRefs extends Component {
  static propTypes = {
    /**
     * Array of refs.
     * Each ref can either be a [callback ref][callback-refs] or an object created with [`React.createRef`][create-ref].
     * Falsy refs are ignored.
     */
    refs: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    ).isRequired,

    /**
     * _Parameters_: `ref: Function`
     *
     * Render the component.
     * `ref` is either a callback ref or `null` if `props.refs` is empty or only contain falsy values.
     */
    render: PropTypes.func.isRequired,
  };

  state = {
    refs: [],
  };

  static getDerivedStateFromProps(props, state) {
    return {
      refs: props.refs.filter(Boolean),
    };
  }

  mergeRefs = ref => {
    this.state.refs.forEach(handler => {
      setRef(ref, handler);
    });
  };

  render() {
    // Don't pass the ref function if there is no refs.
    return this.props.render(
      this.state.refs.length > 0 ? this.mergeRefs : null,
    );
  }
}

export default MergeRefs;
