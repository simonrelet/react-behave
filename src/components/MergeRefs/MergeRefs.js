import PropTypes from 'prop-types'
import React from 'react'
import setRef from './setRef'

/**
 * [callback-refs]: https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
 * [create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref
 *
 * Merge multiple refs on a single component.
 *
 * ## Usage
 *
 * ```jsx
 * import React from 'react'
 * import { ComponentNeedsRef } from 'some-lib'
 * import { MergeRefs } from 'react-behave'
 *
 * class App extends React.Component {
 *   buttonRef = React.createRef()
 *
 *   componentDidMount() {
 *     this.buttonRef.focus()
 *   }
 *
 *   render() {
 *     return (
 *       <ComponentNeedsRef>
 *         {requiredRef => (
 *           <MergeRefs refs={[requiredRef, this.buttonRef]}>
 *             {ref => <button ref={ref}>Click me</button>}
 *           </MergeRefs>
 *         )}
 *       </ComponentNeedsRef>
 *     )
 *   }
 * }
 * ```
 *
 * This component is an alternative of wrapping `<div />`s when you need to have multiple refs on a single component.
 */
class MergeRefs extends React.Component {
  mergeRefs = ref => {
    this.props.refs.forEach(handler => {
      setRef(ref, handler)
    })
  }

  render() {
    return this.props.children(this.mergeRefs)
  }
}

MergeRefs.propTypes = {
  /**
   * _Parameters_: `ref: Function`
   *
   * Render the component.
   */
  children: PropTypes.func.isRequired,

  /**
   * Array of refs.
   * Each ref can either be a [callback ref][callback-refs] or an object created with [`React.createRef`][create-ref].
   */
  refs: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  ).isRequired,
}

export default MergeRefs
