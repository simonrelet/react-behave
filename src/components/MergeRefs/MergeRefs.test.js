import React from 'react'
import { render } from 'react-testing-library'
import MergeRefs from './MergeRefs'

function createProps(props) {
  return {
    children: jest.fn(({ ref }) => <p ref={ref}>Hello</p>),
    refs: [],
    ...props,
  }
}

describe('<MergeRefs />', () => {
  it('renders its content', () => {
    const props = createProps()
    const { container } = render(<MergeRefs {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('merges refs', () => {
    const props = createProps({ refs: [React.createRef(), jest.fn()] })
    const { container } = render(<MergeRefs {...props} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(props.refs[0].current).toEqual(container.firstChild)
    expect(props.refs[1]).toHaveBeenCalledWith(container.firstChild)
  })
})
