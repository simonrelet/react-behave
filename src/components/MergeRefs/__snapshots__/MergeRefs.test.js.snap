// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MergeRefs /> merges refs 1`] = `
<p>
  Hello
</p>
`;

exports[`<MergeRefs /> renders its content 1`] = `
<p>
  Hello
</p>
`;
